File: ca_helpers.php
<?php
defined('MOODLE_INTERNAL') || die();

// Certificates
define('SPE_CA_PUB_B64',         'iuEnM_VNd8OJn9Mxcz4aEp97gIJMAG41nk2CWxj0XhU');
define('SPE_CLIENT_PRIV_B64',    'cSnSlDfvltCuuFqVh80X_0nWwFKLM66rUXIaiIZNqN8');
define('SPE_CLIENT_CERT_JSON',   '{"id":"spe-plugin","pubkey":"NVoaif7KOv-TJ3Flj4vy5gxlYaH2pjS0xrCNlu-qU-I","exp":1793700137,"iss":"SPE-CA"}');
define('SPE_CLIENT_CERT_SIG',    'riqgxsBDCRLtFL4VbWNiS-S0rczoRWsKCdDyUoZuL5GT5DfCGdkNngOAEJdZJpzx-Qa2DZFx6d0FBo21uMCnBQ');
define('SPE_SERVER_CERT_JSON',   '{"id":"spe-api","pubkey":"avOww2r53iLjDacbHGoybcqO10eHw4MUOOaTapiCePA","exp":1793700137,"iss":"SPE-CA"}');
define('SPE_SERVER_CERT_SIG',    '0f75Xhupt6gQYc_mULQHK8JdoUYLxm6wt_u01NYMg726E8-eB4kU2SSGejnwFcDmi8DCMKIgCxJTAlhfGHkACQ');

// Helpers 
function spe_b64u_decode($s) {
    $pad = strlen($s) % 4;
    if ($pad) $s .= str_repeat('=', 4 - $pad);
    return base64_decode(strtr($s, '-_', '+/'));
}
function spe_b64u_encode($b) {
    return rtrim(strtr(base64_encode($b), '+/', '-_'), '=');
}
function spe_lower_header_keys(array $headers): array {
    $norm = [];
    foreach ($headers as $k => $v) {
        $norm[strtolower($k)] = $v;
    }
    return $norm;
}
function spe_normalize_path($path) {
    return $path === '/' ? '/' : rtrim($path, '/');
}

// Client request headers
function spe_ca_build_request_headers(string $path, string $body): array {
    $path = spe_normalize_path($path);

    $cert = json_decode(SPE_CLIENT_CERT_JSON, true);
    if (!$cert || empty($cert['pubkey'])) {
        throw new moodle_exception('Malformed client certificate JSON.');
    }
    if (($cert['exp'] ?? 0) < time()) {
        throw new moodle_exception('Client certificate expired.');
    }
    if (($cert['iss'] ?? '') !== 'SPE-CA' || ($cert['id'] ?? '') !== 'spe-plugin') {
        throw new moodle_exception('Unexpected client certificate claims.');
    }

    $priv_seed = spe_b64u_decode(SPE_CLIENT_PRIV_B64);
    if (strlen($priv_seed) !== 32) {
        throw new moodle_exception('Invalid client private seed length.');
    }
    $pub_raw = spe_b64u_decode($cert['pubkey']);
    if (strlen($pub_raw) !== 32) {
        throw new moodle_exception('Invalid client public key length.');
    }
    $secret_key = $priv_seed . $pub_raw;

    $msg = $path . "\n" . $body;
    $sig = sodium_crypto_sign_detached($msg, $secret_key);

    return [
        'X-SPE-Client-Cert'    => SPE_CLIENT_CERT_JSON,
        'X-SPE-Client-CertSig' => SPE_CLIENT_CERT_SIG,
        'X-SPE-Client-Sig'     => spe_b64u_encode($sig),
    ];
}

// Verify server response
function spe_ca_verify_server_response(string $path, string $body, array $headers): void {
    $path    = spe_normalize_path($path);
    $headers = spe_lower_header_keys($headers);

    $cert_json = $headers['x-spe-server-cert'][0]    ?? $headers['x-spe-server-cert']    ?? '';
    $cert_sig  = $headers['x-spe-server-certsig'][0] ?? $headers['x-spe-server-certsig'] ?? '';
    $srv_sig   = $headers['x-spe-server-sig'][0]     ?? $headers['x-spe-server-sig']     ?? '';

    if (!$cert_json || !$cert_sig || !$srv_sig) {
        throw new moodle_exception('Missing server CA headers.');
    }

    // Verify server certificate signature with CA public key
    $ok = sodium_crypto_sign_verify_detached(
        spe_b64u_decode($cert_sig),
        $cert_json,
        spe_b64u_decode(SPE_CA_PUB_B64)
    );
    if (!$ok) {
        throw new moodle_exception('Invalid server certificate signature.');
    }

    $cert = json_decode($cert_json, true);
    if (!$cert || !isset($cert['pubkey'])) {
        throw new moodle_exception('Malformed server certificate.');
    }
    if (($cert['exp'] ?? 0) < time()) {
        throw new moodle_exception('Server certificate expired.');
    }
    if (($cert['iss'] ?? '') !== 'SPE-CA') {
        throw new moodle_exception('Unexpected server certificate issuer.');
    }
    if (($cert['id'] ?? '') !== 'spe-api') {
        throw new moodle_exception('Unexpected server certificate id.');
    }

    // Verify response signature 
    $msg = $path . "\n" . $body;
    $ok = sodium_crypto_sign_verify_detached(
        spe_b64u_decode($srv_sig),
        $msg,
        spe_b64u_decode($cert['pubkey'])
    );
    if (!$ok) {
        throw new moodle_exception('Invalid server response signature.');
    }
}

File: api/ca_helpers.py

import base64, json, time
from typing import Dict, Any, Mapping
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey, Ed25519PublicKey
)
from cryptography.exceptions import InvalidSignature

# CA public key and server private key / certificate
CA_PUB_B64       = "iuEnM_VNd8OJn9Mxcz4aEp97gIJMAG41nk2CWxj0XhU"
SERVER_PRIV_B64  = "2SWSDhQL-Pw1ayFGcnwJbaOE52-HYQlEIC9dvRL_IUU"
SERVER_CERT_JSON = '{"id":"spe-api","pubkey":"avOww2r53iLjDacbHGoybcqO10eHw4MUOOaTapiCePA","exp":1793700137,"iss":"SPE-CA"}'
SERVER_CERT_SIG  = "0f75Xhupt6gQYc_mULQHK8JdoUYLxm6wt_u01NYMg726E8-eB4kU2SSGejnwFcDmi8DCMKIgCxJTAlhfGHkACQ"

def b64u_decode(s: str) -> bytes:
    s = s + "=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s.encode())

def b64u_encode(b: bytes) -> str:
    return base64.urlsafe_b64encode(b).rstrip(b"=").decode()

def verify_signature(pubkey_b64: str, msg: bytes, sig_b64: str) -> bool:
    try:
        Ed25519PublicKey.from_public_bytes(b64u_decode(pubkey_b64)).verify(
            b64u_decode(sig_b64), msg
        )
        return True
    except InvalidSignature:
        return False

def _lower_headers(h: Mapping[str, str]) -> Dict[str, str]:
    return {str(k).lower(): v for k, v in h.items()}

# Verify client certificate signed by CA
def verify_client_cert(cert_json: str, cert_sig: str) -> Dict[str, Any]:
    """Verify the client (Moodle plugin) certificate using CA public key."""
    ca_pub = Ed25519PublicKey.from_public_bytes(b64u_decode(CA_PUB_B64))
    try:
        ca_pub.verify(b64u_decode(cert_sig), cert_json.encode())
    except InvalidSignature as e:
        raise ValueError("Invalid client certificate signature.") from e
    cert = json.loads(cert_json)
    now = int(time.time())
    if cert.get("exp", 0) < now:
        raise ValueError("Client certificate expired.")
    if cert.get("iss") != "SPE-CA":
        raise ValueError("Unexpected client certificate issuer.")
    if not cert.get("pubkey"):
        raise ValueError("Client certificate missing pubkey.")
    return cert

# Sign server response
def sign_response(path: str, body: str) -> Dict[str, str]:
    """
    Sign server response over the canonical message: f"{path}\\n{body}".
    'path' must be exactly what the client uses when verifying (no trailing slash differences).
    """
    sk = Ed25519PrivateKey.from_private_bytes(b64u_decode(SERVER_PRIV_B64))
    msg = f"{path}\n{body}".encode()
    sig = sk.sign(msg)
    return {
        "X-SPE-Server-Cert": SERVER_CERT_JSON,
        "X-SPE-Server-Sig": b64u_encode(sig),
        "X-SPE-Server-CertSig": SERVER_CERT_SIG,
    }

# Verify client request
def verify_client_request(path: str, body: str, headers: Mapping[str, str]) -> None:
    """
    Verify client request signature and cert chain.
    Accepts any mapping; header names are normalized to lowercase.
    """
    h = _lower_headers(headers)
    cert_json = h.get("x-spe-client-cert")
    cert_sig  = h.get("x-spe-client-certsig")
    req_sig   = h.get("x-spe-client-sig")

    if not cert_json or not cert_sig or not req_sig:
        raise ValueError("Missing client certificate or signature headers.")

    cert = verify_client_cert(cert_json, cert_sig)

    msg = f"{path}\n{body}".encode()
    pubkey_b64 = cert["pubkey"]
    if not verify_signature(pubkey_b64, msg, req_sig):
        raise ValueError("Invalid client request signature.")
